pkg/controller/atlasprivateendpoint/privateendpoint.go

```diff
@@ -0,0 +1,220 @@
+)
+
+func (r *AtlasPrivateEndpointReconciler) handlePrivateEndpointService(
+	ctx *workflow.Context,
+	projectID string,
+	akoPrivateEndpoint *akov2.AtlasPrivateEndpoint,
+) (ctrl.Result, error) {
+	akoPEService := privateendpoint.NewPrivateEndpoint(akoPrivateEndpoint)
+	atlasPEService, err := r.getPrivateEndpointService(ctx.Context, projectID, akoPrivateEndpoint)
+	if err != nil {
+		return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.ReadyType, workflow.Internal, err)
+	}
+
+	wasDeleted := !akoPrivateEndpoint.GetDeletionTimestamp().IsZero()
+	existInAtlas := atlasPEService != nil
+
+	switch {
+	case !existInAtlas && !wasDeleted:
+		atlasPEService, err = r.privateEndpointService.CreatePrivateEndpointService(ctx.Context, projectID, akoPEService)
+		if err != nil {
+			return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceFailedToCreate, err)
+		}
+	case !existInAtlas && wasDeleted:
+		return r.unmanage(ctx, akoPrivateEndpoint)
+	case existInAtlas && wasDeleted:
+		if customresource.IsResourcePolicyKeepOrDefault(akoPrivateEndpoint, r.ObjectDeletionProtection) {
+			return r.unmanage(ctx, akoPrivateEndpoint)
+		}
+
+		atlasPEService, err = r.deletePrivateEndpoint(ctx.Context, projectID, atlasPEService)
+		if err != nil {
+			return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointFailedToDelete, err)
+		}
+	}
+
+	switch atlasPEService.Status() {
+	case privateendpoint.StatusInitiating:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceInitializing, "Private Endpoint is being initialized")
+	case privateendpoint.StatusPending, privateendpoint.StatusPendingAcceptance, privateendpoint.StatusWaitingForUser, privateendpoint.StatusVerified:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceInitializing, "Private Endpoint is waiting for human action")
+	case privateendpoint.StatusFailed, privateendpoint.StatusRejected:
+		return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceFailedToConfigure, errors.New(atlasPEService.ErrorMessage()))
+	case privateendpoint.StatusDeleting:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceDeleting, "Private Endpoint is being deleted")
+	}
+
+	ctx.SetConditionTrue(api.PrivateEndpointServiceReady)
+	r.EventRecorder.Event(akoPrivateEndpoint, "Normal", string(workflow.PrivateEndpointServiceCreated), "Private Endpoint Service is available")
+
+	return r.handlePrivateEndpointInterface(ctx, projectID, akoPrivateEndpoint, akoPEService, atlasPEService)
+}
+
+func (r *AtlasPrivateEndpointReconciler) handlePrivateEndpointInterface(
```

Comment:
The prefix **handle** would made me thing we have another state machine here, but this is more like a **sync**, missing endpoints get created and endpoints not in the spec get removed.

Reply:
Sorry, but I don't get where these assumptions come from. Am I missing some agreement on vocabulary?

Reply:
Maybe it is just me then